ext/sodium/libsodium.stub.php

```diff
@@ -0,0 +1,194 @@
+
+function sodium_crypto_kx_secretkey(string $key): string {}
+
+function sodium_crypto_kx_seed_keypair(string $string): string {}
+
+function sodium_crypto_kx_client_session_keys(string $client_keypair, string $server_key): array {}
+
+function sodium_crypto_kx_server_session_keys(string $server_keypair, string $client_key): array {}
+
+function sodium_crypto_generichash(string $string, string $key = UNKNOWN, int $length = UNKOWN): string {}
+
+function sodium_crypto_generichash_keygen(): string {}
+
+function sodium_crypto_generichash_init(string $key = UNKNOWN, int $length = UNKNOWN): string {}
+
+function sodium_crypto_generichash_update(string &$state, string $string): bool {}
+
+function sodium_crypto_generichash_final(string &$state, $length = UNKNOWN): string {}
+
+function sodium_crypto_kdf_derive_from_key(int $subkey_len, int $subkey_id, string $context, string $key): string {}
+
+function sodium_crypto_kdf_keygen(): string {}
+
+#ifdef crypto_pwhash_SALTBYTES
+function sodium_crypto_pwhash(int $length, string $password, string $salt, int $opslimit, int $memlimit, int $alg = UNKNOWN): string {}
+
+function sodium_crypto_pwhash_str(string $password, int $opslimit, int $memlimit): string {}
+
+function sodium_crypto_pwhash_str_verify(string $hash, string $password): bool {}
+#endif
+
+#if SODIUM_LIBRARY_VERSION_MAJOR > 9 || (SODIUM_LIBRARY_VERSION_MAJOR == 9 && SODIUM_LIBRARY_VERSION_MINOR >= 6)
+function sodium_crypto_pwhash_str_needs_rehash(string $password, int $opslimit, int $memlimit): bool {}
+#endif
+
+#ifdef crypto_pwhash_scryptsalsa208sha256_SALTBYTES
+function sodium_crypto_pwhash_scryptsalsa208sha256(int $length, string $password, string $salt, int $opslimit, int $memlimit): string {}
+
+function sodium_crypto_pwhash_scryptsalsa208sha256_str(string $password, int $opslimit, int $memlimit): string {}
+
+function sodium_crypto_pwhash_scryptsalsa208sha256_str_verify(string $hash, string $password): bool {}
+#endif
+
+function sodium_crypto_scalarmult(string $string_1, string $string_2): string {}
+
+function sodium_crypto_secretbox(string $string, string $nonce, string $key): string {}
+
+function sodium_crypto_secretbox_keygen(): string {}
+
+/** @return string|false */
+function sodium_crypto_secretbox_open(string $string, string $nonce, string $key) {}
+
+#ifdef crypto_secretstream_xchacha20poly1305_ABYTES
+function sodium_crypto_secretstream_xchacha20poly1305_keygen(): string {}
+
+function sodium_crypto_secretstream_xchacha20poly1305_init_push(string $key): array {}
+
+function sodium_crypto_secretstream_xchacha20poly1305_push(string &$state, string $string_1, string $string_2 = UNKNOWN, int $long = UNKNOWN): string {}
+
+function sodium_crypto_secretstream_xchacha20poly1305_init_pull(string $string, string $key): string {}
+
+/** @return array|false */
+function sodium_crypto_secretstream_xchacha20poly1305_pull(string &$state, string $string_1, string $string_2 = UNKNOWN) {}
```

Comment:
These functions check that the ref arguments is a string and throw `SodiumException` in that case. I guess our contract here is just that if the type is violated we should throw *something*, it needn't be TypeError in particular, right?

Reply:
IMHO, it doesn't really matter which exception is thrown if anything else than a string is given. Of course, changing the exception type constitutes a BC break, though. It might make sense to not specify the type here (for now), since otherwise the implementation should be adapted as well.